engine/verification/assigner/engine.go

```diff
@@ -63,36 +73,159 @@ func (e *Engine) Done() <-chan struct{} {
+
+	// pushes assigned chunks to chunks queue.
+	e.processChunks(chunkList, resultID)
 }
 
-func (e *Engine) Submit(originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(originID, event)
+// processChunks receives a list of chunks all belong to the same execution result. It creates a chunk locator
+// for each of those chunks and stores the chunk locator in the chunks queue.
+//
+// Note that all chunks in the input chunk list are assume to be legitimately assigned to this verification node
+// (through the chunk assigner), and all belong to the same execution result.
+//
+// Deduplication of chunk locators is delegated to the chunks queue.
+func (e *Engine) processChunks(chunkList flow.ChunkList, resultID flow.Identifier) {
+	for _, chunk := range chunkList {
+		log := e.log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Hex("chunk_id", logging.ID(chunk.ID())).
+			Uint64("chunk_index", chunk.Index).Logger()
+
+		locator := &chunks.Locator{
+			ResultID: resultID,
+			Index:    chunk.Index,
+		}
+
+		// pushes chunk locator to the chunks queue
+		ok, err := e.chunksQueue.StoreChunkLocator(locator)
 		if err != nil {
-			engine.LogError(e.log, err)
+			log.Error().Err(err).Msg("could not push chunk locator to chunks queue")
+			continue
+		}
+		if !ok {
+			log.Debug().Msg("could not push duplicate chunk locator to chunks queue")
+			continue
 		}
-	})
+
+		// notifies chunk queue consumer of a new chunk
+		e.newChunkListener.Check()
+		log.Debug().Msg("chunk locator successfully pushed to chunks queue")
+	}
 }
 
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.Process(e.me.NodeID(), event)
+// stakedAtBlockID checks whether this instance of verification node has staked at specified block ID.
+// It returns true and nil if verification node is staked at referenced block ID, and returns false and nil otherwise.
+// It returns false and error if it could not extract the stake of node as a verification node at the specified block.
+func (e *Engine) stakedAtBlockID(blockID flow.Identifier) (bool, error) {
+	identity, err := protocol.IdentityAtBlockID(e.state, blockID, e.me.NodeID())
+	if err != nil {
+		return false, fmt.Errorf("could not extract staked identify of node at block %v: %w", blockID, err)
+	}
+
+	if identity.Role != flow.RoleVerification {
+		return false, fmt.Errorf("node is staked for an invalid role. expected: %s, got: %s", flow.RoleVerification, identity.Role)
+	}
+
+	staked := identity.Stake > 0
+	return staked, nil
 }
 
-func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+// ProcessFinalizedBlock indexes the execution receipts included in the block, and handles their chunk assignments.
+// Once it is done handling all the receipts in the block, it notifies the block consumer.
+func (e *Engine) ProcessFinalizedBlock(block *flow.Block) {
```

Comment:
Not related to this line, but we need to implement the `FinalizedBlockReader` in the assigner/consumer.go

Reply:
Issued up for this sprint https://github.com/dapperlabs/flow-go/issues/5353